extensions/amp-story/1.0/amp-story-embedded-component.js

```diff
@@ -0,0 +1,606 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  Action,
+  StateProperty,
+  UIType,
+  getStoreService,
+} from './amp-story-store-service';
+import {CSS} from '../../../build/amp-story-tooltip-1.0.css';
+import {EventType, dispatch} from './events';
+import {Services} from '../../../src/services';
+import {addAttributesToElement, closest, matches} from '../../../src/dom';
+import {createShadowRootWithStyle, getSourceOriginForElement} from './utils';
+import {dev, devAssert, user} from '../../../src/log';
+import {dict} from '../../../src/utils/object';
+import {getAmpdoc} from '../../../src/service';
+import {htmlFor, htmlRefs} from '../../../src/static-template';
+import {isProtocolValid, parseUrlDeprecated} from '../../../src/url';
+import {setImportantStyles, toggle} from '../../../src/style';
+
+/**
+ * Enum of elements that can be expanded.
+ * @enum {string}
+ */
+const EXPANDABLE_COMPONENTS = {
+  TWITTER: 'amp-twitter',
+};
+
+/**
+ * Enum containing all interactive components.
+ * @enum {string}
+ */
+const interactiveComponents = Object.assign({}, EXPANDABLE_COMPONENTS, {
+  EXPANDED_VIEW_OVERLAY: '.i-amphtml-story-expanded-view-overflow, ' +
+    '.i-amphtml-expanded-view-close-button',
+  LINK: 'a[href]',
+});
+
+/**
+ * Selectors that should delegate to AmpStoryEmbeddedComponent.
+ */
+export function embeddedComponentSelectors() {
+  // Using indirect invocation to prevent no-export-side-effect issue.
+  return interactiveComponents;
+}
+
+/**
+ * Builds expanded view overlay for expandable components.
+ * @param {!Element} element
+ * @return {!Element}
+ */
+const buildExpandedViewOverlay = element => htmlFor(element)`
+    <div class="i-amphtml-story-expanded-view-overflow
+        i-amphtml-story-system-reset">
+      <span class="i-amphtml-expanded-view-close-button" role="button">
+        &times;
+      </span>
+    </div>`;
+
+/**
+ * Minimum vertical space needed to position tooltip.
+ * @const {number}
+ */
+const MIN_VERTICAL_SPACE = 56;
+
+/**
+ * Padding between tooltip and edges of screen.
+ * @const {number}
+ */
+const EDGE_PADDING = 8;
+
+/**
+ * Blank icon when no data-tooltip-icon src is specified.
+ * @const {string}
+ */
+const DEFAULT_ICON_SRC =
+  'data:image/gif;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs=';
+
+/**
+ * @struct @typedef {{
+ *   tooltip: !Element,
+ *   buttonLeft: !Element,
+ *   buttonRight: !Element,
+ *   arrow: !Element,
+ * }}
+ */
+let tooltipElementsDef;
+
+const TAG = 'amp-story-embedded-component';
+
+const ComponentState = {
```

Comment:
Nice!!

Nit: here and above, please add some jsdoc

Reply:
Done